File: problem_solving/hash_table/py/49_group_anagrams.py
'''
https://leetcode.com/problems/group-anagrams/description/?envType=study-plan-v2&envId=top-interview-150

create frequency counter... 
dict[tuple of freq. counter] = Lis

'''
from collections import defaultdict

#O(m * n)
def group_anagrams(strs:list[str])->list[list[str]]:
    result_hash = {}
    for st in strs:
        count = [0]*26
        for s in st:
            count[ord(s) - ord('a')] += 1    
        t = tuple(count)        #we need tuple as key
        if t in result_hash:
            result_hash[t].append(st) 
        else:
            result_hash[t] = [st]

    return list(result_hash.values())   #mistake ... need to change this to list


strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
result = group_anagrams(strs)
print("result: " + str(result))


strs = [""]
result = group_anagrams(strs)
print("result: " + str(result))

File: problem_solving/hash_table/cpp/49_group_anagrams.cpp
/*
https://leetcode.com/problems/group-anagrams/description/?envType=study-plan-v2&envId=top-interview-150

std::unordered_map<std::string, std::vector<std::string>> 
create frequency counter... 
create string key with count#count#...

'''
*/
#include<iostream>
#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"
#include<unordered_map>
using namespace std;

std::vector<std::vector<string>> group_anagrams(const std::vector<std::string>& strs ) {
    
    std::unordered_map<std::string, std::vector<string>> map;
    for(const auto& str : strs) {
        std::vector<int> key(26, 0);
        for(const auto& ch : str) {
            key[ch - 'a']++; 
        }
        std::string key_str;
        for(int i : key) {
            key_str += std::to_string(i) + "#";
        }
        std::cout << key_str << std::endl;
        map[key_str].push_back(str);
    }

    std::vector<std::vector<string>> result;
    for(const auto& pair : map) {
        result.push_back(pair.second);
    }
    return result;
}




int main() {
    std::vector<std::string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
    std::vector<std::vector<std::string>> result = group_anagrams(strs);
    print_vector_of_vector(result);
}


